provider/docker/event/listener.go

```diff
@@ -0,0 +1,130 @@
+package event
+
+import (
+	"context"
+	"time"
+
+	"github.com/containous/traefik/log"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/events"
+	"github.com/docker/docker/api/types/versions"
+	"github.com/docker/docker/client"
+)
+
+const (
+	// SwarmDefaultWatchTime is the duration of the interval when polling docker
+	SwarmDefaultWatchTime = 15 * time.Second
+)
+
+// NewListener creates a new event listener depending on what the user's Docker daemon supports.
+func NewListener(dockerClient client.APIClient, dockerEventsOptions types.EventsOptions, stopChan chan bool, errChan chan error, callback Callback) (Listener, error) {
+	serverVersion, err := dockerClient.ServerVersion(context.Background())
+	if err != nil {
+		return nil, err
+	}
+
+	// https://docs.docker.com/engine/api/v1.29/#tag/Network (Docker 17.06)
+	if versions.GreaterThanOrEqualTo(serverVersion.APIVersion, "1.29") {
+		eventsCtx := context.Background()
+		eventsMsgChan, eventsErrChan := dockerClient.Events(
+			eventsCtx,
+			dockerEventsOptions,
+		)
+
+		e := &Streamer{
+			EventsMsgChan: eventsMsgChan,
+			EventsErrChan: eventsErrChan,
+			Callback:      callback,
+			StopChan:      stopChan,
+			ErrChan:       errChan,
+			EventsCtx:     eventsCtx,
+		}
+
+		return e, nil
+	}
+
+	// Fallback to the ticker.
+	e := &Ticker{
+		Callback:       callback,
+		StopChan:       stopChan,
+		TickerInterval: SwarmDefaultWatchTime,
+	}
+
+	return e, nil
+}
+
+// Listener is an interface for event listeners to implement.
+type Listener interface {
+	Start()
+	Stop()
```

Comment:
This is never called from outside the package, shouldn't be defined here (should just be handled internally by the several implementations)